### ISSUE TITLE
Push notifications

### ISSUE THREAD

**Issue by plmercereau** [nhost/hasura-auth#239]
@L3K0V suggested to add a push notification mechanism. This issue is for tracking the idea and possible solution.
Could you explain more what you have in mind, @L3K0V ?

**Comment by L3K0V** [comment]
Sure!

First of all I'm having a mobile-first native apps and the plan was to use Firebase push notifications for both Android and iOS.

I have introduced a concept of installations in my nhost stack maintaining a table and device installations. An installation have some information about the device like platform version, app identifier or bundle id etc. and most important - a device token.

Using a serverless functions and these installations I'm querying users devices and their token to send notifications to.

It will be interesting how others are implementing a push notifications 🤔 

@plmercereau do you think a concept of installations will be something for hasura-auth with some adapter approach supporting different push notification providers. Maybe I can open some PR or at first discussion.

**Comment by elitan** [comment]
Adding docs from Firebase for reference: https://firebase.google.com/docs/cloud-messaging

**Comment by plmercereau** [comment]
Thinking out loud. What if:
- (optional) we set a table `auth.user_devices` (id, user_id, device_token)
- every time hasura-auth generates a jwt, we embed the device token in the JWT (`x-hasura-device-token`)
- we set a table `notifications` (id, message_id, device_token, read `boolean default false`
  - set `select` permission to `{ device_token: { _eq: "x-hasura-device-token" }, read: { _eq: false } }`
  - set a similar `update` permission, with a preset on `read` to `true`
- we set a `messages` (id, user_id, body) with an insert trigger: "insert new notifications for this message and all the devices of the user"
  - set `select` permission to `{ user_id: { _eq: "x-hasura-user-id" } }`


Then on the client side:
```graphql
subscription listenToSubscriptions {
  notifications {
    message {
      id
      body
    }
  }
}
```

Every time a new subscription is received, we run a mutation:
```graphql
mutation markAsRead($id: uuid!)  {
  set_notification($id) 
}
```

A way to start smoothly would be to find a consistent way to determine a consistent device token, either from the frontend or from the http headers (not expert on the subject), and to add it to the jwt. The rest doesn't necessary need to be part of hasura-auth or a "core" Nhost product.
